### PR TITLE
Audit: track user events

### PR DESCRIPTION
**Pull request by stsewd**
This is to allow us to offer users some security logs around log in users and page views.

I realized that some events can be attached to a user, and others could also be attached to a project (so we are able to filter events per project, like page views). I'm not sure if we should keep these after the attached project is deleted, maybe we do?

Another thing is that creating these events on doc serving will trigger a lot of other resources and becomes quite noisy
Screenshot 2021-07-28 at 18-10-37 Select audit log to change Django site admin

So, I think it makes sense to attach this event to the analytics endpoint, but a downside is that someone can skip that easily.

This is missing tests, but wanted to confirm those points before writing them

This is another screenshot using another browser

Screenshot 2021-07-28 at 18-10-48 Select audit log to change Django site admin

### CODE REVIEW

**Comment by ericholscher** [comment]
@stsewd I think we want it attached to proxito still for auditing, but we don't want to record media queries. I think we can filter by HTML & PDF, which should be enough. I'm guessing we already have this logic somewhere in the codebase I imagine, but `*.html` might do the trick for the basic version. Eventually we could make the file extensions configurable perhaps.

**Comment by stsewd** [comment]
@ericholscher that did the trick! What about my other question, should we delete the logs attached to a project (page views for now) after a project is deleted? or keep them the same way when a user is deleted?

**Comment by ericholscher** [comment]
I think we want to keep everything if possible. So probably a similar approach to the user. For auditing we definitely don't want people to be able to cover their tracks by deleting something they have access to.

**Comment by stsewd** [comment]
ok, I think another step would be to have a task to clean those after some time, pageviews will generate a lot of records.

**Comment by ericholscher** [comment]
@stsewd yea -- we're only going to turn this on for specific paid plans, not everyone. Part of what they'll be paying for is storage. If it becomes super large, we could always record these in a secondary DB, but we're a long way off from that :)

File: readthedocs/proxito/views/mixins.py [hunk 1]
```diff
@@ -68,6 +71,32 @@ def _serve_docs(
             download=download,
         )
 
+    def _trak_pageview(self, project, path, request):
```
Comment (ericholscher) [inline readthedocs/proxito/views/mixins.py:74]:
```suggestion
    def _track_pageview(self, project, path, request):
```

File: readthedocs/proxito/views/mixins.py [hunk 2]
```diff
@@ -68,6 +71,32 @@ def _serve_docs(
...
+            )
+
+    def _is_audit_enabled(self, project):
+        """
+        Check if the project has the audit feature enabled to track individual page views.
+
+        This feature is different from page views analytics,
+        as it records every page view individually with more metadata like the user, IP, etc.
+        """
+        return True
```
Comment (ericholscher) [inline readthedocs/proxito/views/mixins.py:98]: Probably default this to `False` I think. No reason to have it on unless we really want it to be.

File: readthedocs/acl/utils.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+from django.contrib.auth import BACKEND_SESSION_KEY
+
+from readthedocs.acl.constants import BACKEND_REQUEST_KEY
+
+
+def get_auth_backend(request):
```
Comment (ericholscher) [inline readthedocs/acl/utils.py:6]: I feel like we already have this code on .com -- this is porting it right?

File: readthedocs/proxito/views/mixins.py [hunk 4]
```diff
@@ -68,6 +71,32 @@ def _serve_docs(
             download=download,
         )
 
+    def _trak_pageview(self, project, path, request):
+        """Create an audit log of the page view if audit is enabled."""
+        # Remove any query args (like the token access from AWS).
+        path_only = urlparse(path).path
+        track_file = any(
+            path_only.endswith(ext)
+            for ext in ['.html', '.pdf', '.epub', '.zip']
```
Comment (ericholscher) [inline readthedocs/proxito/views/mixins.py:80]: I think `endswith` accepts a list to look for: https://docs.python.org/3/library/stdtypes.html#str.endswith

**Review by ericholscher — COMMENTED** [review]
This looks like a good start 👍 Excited to get the login/logout audit log, and then I think we have the basics ready to ship and test out, along with billing integration on .com

File: readthedocs/acl/utils.py (hunk 3 above)
Comment (stsewd) [reply readthedocs/acl/utils.py:6]: yeah, I'm porting this piece from .com, so we have all code here.

File: readthedocs/proxito/views/mixins.py (hunk 2 above)
Comment (stsewd) [reply readthedocs/proxito/views/mixins.py:98]: yeah, I have it as true for testing for now.

File: readthedocs/proxito/views/mixins.py (hunk 4 above)
Comment (stsewd) [reply readthedocs/proxito/views/mixins.py:80]: Nice!

**Comment by stsewd** [comment]
This should be ready for review now, I'm only missing writing some tests on .com.

**Review by ericholscher — APPROVED** [review]
Changes look good. Curious what our plan for rolling this out is. I'd like to figure out a way to turn it on for one of our projects first, so maybe we can do that on .com. Currently there's no way to enable it on .org -- should we add a feature flag or something simple to enable it for testing?

**Comment by stsewd** [comment]
> should we add a feature flag or something simple to enable it for testing?

I was thinking that maybe we can find a way to have a mapping of some features flags to plan features, so we don't have to change our code in .org/.com or leave it as a method that only returns true/false every time we have to do a check like this, but only check for `project.has_feature(Feature)` and check for feature flags and the plans mapping internally.

But the real testing here is when using the auth backends on page views, log in and log out are enabled on both sites (we could also kill it on .org if we consider)

**Comment by ericholscher** [comment]
Sounds good. I think we can move forward 👍

**Pull request merged**